core/src/banking_stage/transaction_scheduler/prio_graph_scheduler.rs

```diff
@@ -0,0 +1,660 @@
+use {
+    super::{
+        in_flight_tracker::InFlightTracker,
+        scheduler_error::SchedulerError,
+        thread_aware_account_locks::{ThreadAwareAccountLocks, ThreadId, ThreadSet},
+        transaction_state::SanitizedTransactionTTL,
+        transaction_state_container::TransactionStateContainer,
+    },
+    crate::banking_stage::{
+        consumer::TARGET_NUM_TRANSACTIONS_PER_BATCH,
+        read_write_account_set::ReadWriteAccountSet,
+        scheduler_messages::{ConsumeWork, FinishedConsumeWork, TransactionBatchId, TransactionId},
+        transaction_scheduler::transaction_priority_id::TransactionPriorityId,
+    },
+    crossbeam_channel::{Receiver, Sender, TryRecvError},
+    itertools::izip,
+    prio_graph::{AccessKind, PrioGraph},
+    solana_sdk::{pubkey::Pubkey, slot_history::Slot, transaction::SanitizedTransaction},
+    std::collections::HashMap,
+};
+
+pub(crate) struct PrioGraphScheduler {
+    in_flight_tracker: InFlightTracker,
+    account_locks: ThreadAwareAccountLocks,
+    consume_work_senders: Vec<Sender<ConsumeWork>>,
+    finished_consume_work_receiver: Receiver<FinishedConsumeWork>,
+    look_ahead_window_size: usize,
+}
+
+impl PrioGraphScheduler {
+    pub(crate) fn new(
+        consume_work_senders: Vec<Sender<ConsumeWork>>,
+        finished_consume_work_receiver: Receiver<FinishedConsumeWork>,
+    ) -> Self {
+        let num_threads = consume_work_senders.len();
+        Self {
+            in_flight_tracker: InFlightTracker::new(num_threads),
+            account_locks: ThreadAwareAccountLocks::new(num_threads),
+            consume_work_senders,
+            finished_consume_work_receiver,
+            look_ahead_window_size: 2048,
```

Comment:
I'm trying to hack a simulator together this next week; it won't be as robust as @ryoqun's simulation tool, but I can use it to get some idea on scheduling efficiency by looking at the schedule depth.

i.e. we schedule `(A, B, C, D)` on thread 0, `(E, F, G, H, I, J, K, L, M)` on thread 2 - the depth is 9 vs if we had a more balanced `(A, B, C, D, J, K)` and `(E, F, G, H, I, L, M)` w/ max depth of 7.

This can basically tell us if the look-ahead is causing us to queue too much work on a single thread.